src/webgpu/api/validation/capability_checks/limits/maxFragmentCombinedOutputResources.spec.ts

```diff
@@ -0,0 +1,124 @@
+      return vec4f(0);
+    }
+
+    // these are unused and so should not affect the test.
+    @group(3) @binding(0) var unusedTexture1d: texture_storage_1d<rgba32float, write>;
+    @group(3) @binding(1) var unusedTexture2d: texture_storage_2d<rgba32float, write>;
+    @group(3) @binding(2) var unusedTexture2dArray: texture_storage_2d_array<rgba32float, write>;
+    @group(3) @binding(3) var unusedTexture3d: texture_storage_3d<rgba32float, write>;
+
+    // testValue: ${testValue}
+    // numStorageTextures: ${numStorageResources}
+    // numTargets: ${numTargets}
+
+    // We use rgba32float format storage textures as they take the most size and so are most likely
+    // to trigger any invalid validation.
+
+    ${range(
+      numStorageResources,
+      i => `${bindingAndGroup(i)} ${i % 2 ? textureDecl(i) : bufferDecl(i)}`
+    ).join('\n    ')}
+
+    @fragment fn fs() -> @location(0) vec4f {
+      ${range(numStorageResources, i => (i % 2 ? textureUsage(i) : bufferUsage(i))).join(
+        '\n      '
+      )}
+      return vec4f(0);
+    }
+  `;
+  const module = device.createShaderModule({ code });
+  const pipelineDescriptor: GPURenderPipelineDescriptor = {
+    layout: 'auto',
+    vertex: {
+      module,
+      entryPoint: 'vs',
+    },
+    fragment: {
+      module,
+      entryPoint: 'fs',
+      targets: new Array(numTargets).fill({ format: 'r8unorm', writeMask: 0 }),
+    },
+  };
+  return {
+    pipelineDescriptor,
+    code,
+    numStorageTextures: ((numStorageResources + 1) / 2) | 0,
+    numStorageBuffers: (numStorageResources / 2) | 0,
+  };
+}
+
+const kExtraLimits: LimitsRequest = {
+  maxColorAttachments: 'adapterLimit',
+  maxColorAttachmentBytesPerSample: 'adapterLimit',
+  maxStorageBuffersPerShaderStage: 'adapterLimit',
+  maxStorageTexturesPerShaderStage: 'adapterLimit',
+};
+
+const limit = 'maxFragmentCombinedOutputResources';
+export const { g, description } = makeLimitTestGroup(limit);
+
+g.test('createRenderPipeline,async,at_over')
+  .desc(`Test using at and over ${limit} limit in createRenderPipeline(Async)`)
+  .params(kMaximumLimitBaseParams.combine('async', [false, true] as const))
+  .fn(async t => {
+    const { limitTest, testValueName, async } = t.params;
+    await t.testDeviceWithRequestedMaximumLimits(
+      limitTest,
+      testValueName,
+      async ({ device, testValue, shouldError }) => {
+        const {
+          pipelineDescriptor,
+          code,
+          numStorageBuffers,
+          numStorageTextures,
+        } = getPipelineDescriptor(device, testValue);
+        const targets = pipelineDescriptor.fragment?.targets as GPUColorTargetState[];
+        const bytesPerSample = computeBytesPerSample(targets);
+        if (
+          targets.length > device.limits.maxColorAttachments ||
+          bytesPerSample > device.limits.maxColorAttachmentBytesPerSample ||
+          numStorageBuffers > device.limits.maxStorageBuffersPerShaderStage ||
```

Comment:
Ah... because I know that in the Dawn backend, `maxStorage[Buffer|Texture]PerShaderStage` are both clamped to 8, but `maxFragmentCombinedOutputResources` is computed as sum of `maxStorage[Buffer|Texture]PerShaderStage` and `maxColorAttachment` pre-clamping, I think that this condition might happen more often than ideal... WDYT about clamping the number of storageBuffers and storageTextures to their max limit also in case of skipping? Then worse case scenario we are still testing the effective (allowed) maximum of `maxFragmentCombinedOutputResources`?

Reply:
done